core/ordering/cosipbft/blockstore/mem.go

```diff
@@ -165,23 +158,100 @@ func (s *InMemory) WithTx(txn store.Transaction) BlockStore {
 	return store
 }
 
+// Observer is an observer that can be added to store watcher. It will announce
+// the blocks in order and without blocking the watcher even if the listener is
+// not actively emptying the queue.
+//
+// - implements core.Observer.
 type observer struct {
-	ch chan types.BlockLink
+	sync.Mutex
+
+	buffer  []types.BlockLink
+	running bool
+	closed  bool
+	ch      chan types.BlockLink
+}
+
+func newObserver(ctx context.Context, watcher core.Observable) *observer {
+	ch := make(chan types.BlockLink, 1)
+	obs := &observer{
+		ch: ch,
+	}
+
+	watcher.Add(obs)
+
+	go func() {
+		<-ctx.Done()
+		watcher.Remove(obs)
+		obs.close()
+	}()
+
+	return obs
+}
+
+// NotifyCallback implements core.Observer.
+func (obs *observer) NotifyCallback(evt interface{}) {
+	obs.Lock()
+	defer obs.Unlock()
+
+	if obs.closed {
+		return
+	}
+
+	if obs.running {
+		obs.buffer = append(obs.buffer, evt.(types.BlockLink))
+		return
+	}
+
+	select {
+	case obs.ch <- evt.(types.BlockLink):
+
+	default:
+		// The buffer size is not controlled as we assume the event will be read
+		// shortly by the caller.
```

Comment:
this looks reasonable at the moment, but also like the kind of thing that could some day blow up in our face and might take some time to debug.

Even if we're not handling it, can we at least put some conditional WARN logging if the buffer is above a certain (insane) size?

Reply:
That seems reasonable ;)